スクレイピング処理とレコード作成処理を各Scraperクラスから分離

mterada1228 opened the pull request.

mterada1228 committed:
スクレイピング処理とレコード作成処理を各Scraperクラスから分離

ershn commented on app/services/scraper/creator.rb:
```diff
@@ -0,0 +1,13 @@
+module Scraper
+  class Creator
```
必ずゼロからの生成だけではないので、`Upserter`という名称の方が適切だと思います。

mterada1228 replied:
`Upserter` にクラス名称を変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/parser.rb:
```diff
@@ -0,0 +1,16 @@
+      def parse_horse(response)
+        document = Nokogiri::HTML(response)
+
+        item = Horse.new
```
依存関係を単純化するため、ここで`Horse`モデルを直接使わず、ハッシュを返す方が保守性がいいと思います。

mterada1228 replied:
このクラスではモデルを生成せずに、ハッシュを返すのみに変更しました。  
`HorseScraperService` 内でモデルを生成しています。

ershn replied:
OK

ershn commented on app/services/scraper/parser.rb:
```diff
@@ -0,0 +1,16 @@
+      end
+    end
+  end
+end
```
url の呼び出しはこのファイル内で行ったほうがいいと思います。

そうした上で名称は`Scraper`の方が適切かと。

mterada1228 replied:
`url` を使用する処理は全て `Scraper` クラスにまとめました。

ershn replied:
OK

ershn commented on app/services/scraper/creator.rb:
```diff
@@ -0,0 +1,13 @@
+      end
+    end
+  end
+end
```
クラスメソッドを追加せず、initialize に何も引数を渡さなくてもインスタンスメソッドにしたほうがいいと思います。

mterada1228 replied:
クラスメソッドを廃し、インスタンスメソッドに変更しています。

ershn replied:
OK

ershn commented on app/services/scraper/parser.rb:
```diff
@@ -0,0 +1,16 @@
+      end
+    end
+  end
+end
```
クラスメソッドを追加せず、initialize に何も引数を渡さなくてもインスタンスメソッドにしたほうがいいと思います。

mterada1228 replied:
クラスメソッドを廃し、インスタンスメソッドに変更しています。

ershn replied:
OK

mterada1228 committed:
refs features/devide_class_of_scrapers 指摘対応

ershn commented on app/services/scraper/scraper.rb:
```diff
@@ -0,0 +1,15 @@
+        document.css('#db_main_box > div.db_head.fc > div.db_head_name.fc > div.horse_title > h1')
+                .text.gsub(/[[:space:]]/, '')
+      item
+    end
```
引数は`initialize`に渡したほうが実装がより自由にできるようになります。

mterada1228 replied:
`HorseScraper`, `HorseUpserter` はともに、`initializer` として引数を受け取るように変更し、メソッドの引数は無くしました。

ershn replied:
OK

ershn commented on app/services/scraper/upserter.rb:
```diff
@@ -0,0 +1,11 @@
+      attributes = record.attributes
+      record = record.class.find(record[record.class.primary_key])
+      record.update(attributes)
+    end
```
> 引数は`initialize`に渡したほうが実装がより自由にできるようになります。

mterada1228 replied:
> HorseScraper, HorseUpserter はともに、initializer として引数を受け取るように変更し、メソッドの引数は無くしました。

ershn replied:
OK

ershn commented on app/services/scraper/scraper.rb:
```diff
@@ -0,0 +1,15 @@
+      item
+    end
+  end
+end
```
この処理は`Horse`のためのものなので、モジュールは`Scraper::HorseScraper`などの方が適切だと思います。

mterada1228 replied:
`Scraper::Scraper` ではなく `Scraper::HorseScraper` に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/scraper.rb:
```diff
@@ -0,0 +1,15 @@
+module Scraper
+  class Scraper
+    def scrape_horse(url)
```
処理の意図はクラス名で伝わるので、このメソッド名は`call`など汎用的なものにしても問題ないです。

mterada1228 replied:
メソッド名は `call` に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/upserter.rb:
```diff
@@ -0,0 +1,11 @@
+      attributes = record.attributes
+      record = record.class.find(record[record.class.primary_key])
+      record.update(attributes)
+    end
```
この処理は`find_or_initialize_by`を使うことでより簡単に実装できると思います。

そうすれば、汎用的なクラスの必要もなくなるかもしれません。

mterada1228 replied:
`find_or_initialize_by(id: attributes[:id])` として、primary key で検索して見つかれば`Horse.find`、見つからなければ`Horse.new` される処理にしました。その後で `Horse#update(attributes)` をしてレコードの `insert` または `update` を行うようにしました。

ershn replied:
OK

ershn commented on app/services/scraper/upserter.rb:
```diff
@@ -0,0 +1,11 @@
+module Scraper
+  class Upserter
+    def create(record)
```
> 処理の意図はクラス名で伝わるので、このメソッド名はcallなど汎用的なものにしても問題ないです。

mterada1228 replied:
メソッド名は call に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/horse_scraper_service.rb:
```diff
@@ -10,9 +10,8 @@ def initialize(url:)
-      horse = Scraper::Parser.parse_horse(response)
-      Scraper::Creator.create(horse)
+      attributes = Scraper.new.scrape_horse(@url)
+      Upserter.new.create(Horse.new(attributes))
```
汎用的なクラスじゃなければ、`Horse.new`は`Upserter`内で行った方が役割分担が明白になります。

mterada1228 replied:
モデルに関する処理をまとめるために、`Scraper::HorseUpserter` 内で、`Horse.find_or_initialize_by` => `Horse#update` することで、Horse モデルの生成 /更新をするよう変更しました。

ershn replied:
OK

mterada1228 committed:
指摘対応

ershn commented on app/services/scraper/horse_scraper_service.rb:
```diff
@@ -10,28 +10,9 @@ def initialize(url:)
-      Rails.logger.info(attributes)
-    end
   end
 end
```
`HorseScraper`と`HorseScraperService`の2つのクラスがありますが、名前を見るだけでは何が違うのかがぱっとわからないので、調整しておいた方がいいと思います。

ここは`HorseScraperService`を一つ上のディレクトリに置くことでそのクラスの方が汎用的だと伝えられると思います。

mterada1228 replied:
`Scraper::HorseScraperService` を `HorseScrperService` に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/horse_upserter.rb:
```diff
@@ -0,0 +1,12 @@
+
+    def call
+      horse = Horse.find_or_initialize_by(id: @attributes[:id])
+      horse.update(@attributes)
```
保存が失敗する可能性があるのですが、`update`だと、`false`が返されるだけで、エラーに気づかない可能性があるので、ここは`update!`を使ったほうが安全です。

mterada1228 replied:
`HorseUpserter#call` の `horse.update` を `horse.update!` に変更しました。

ershn replied:
OK

mterada1228 committed:
指摘対応
  - Scraper::HorseScraperService を HorseScrperService に変更
  - HorseUpserter#call の horse.update を horse.update! に変更
